lib/ts/types.ts

```diff
@@ -17,25 +17,26 @@ import RecipeModule from "./recipeModule";
 import NormalisedURLDomain from "./normalisedURLDomain";
 import NormalisedURLPath from "./normalisedURLPath";
 import { TypeFramework } from "./framework/types";
+import { BaseRequest } from "./framework";
 
 export type AppInfo = {
     appName: string;
-    websiteDomain: string;
-    websiteBasePath?: string;
+    origin: string | ((req: BaseRequest, userContext: any) => Promise<string>);
+    originBasePath?: string;
     apiDomain: string;
     apiBasePath?: string;
     apiGatewayPath?: string;
 };
 
 export type NormalisedAppinfo = {
     appName: string;
-    websiteDomain: NormalisedURLDomain;
+    origin: (req: BaseRequest, userContext: any) => Promise<NormalisedURLDomain>;
     apiDomain: NormalisedURLDomain;
     topLevelAPIDomain: string;
-    topLevelWebsiteDomain: string;
     apiBasePath: NormalisedURLPath;
     apiGatewayPath: NormalisedURLPath;
-    websiteBasePath: NormalisedURLPath;
+    originBasePath: NormalisedURLPath;
+    initialOriginType: string;
```

Comment:
Also this is bad typing, if you always expect it to be "string" or "function" type it to that instead of leaving it open ended as a string